execution_engine/src/core/runtime/mod.rs

```diff
@@ -1196,12 +1137,92 @@ where
 
     fn execute_contract(
         &mut self,
-        contract_package: ContractPackage,
-        contract_hash: ContractHash,
-        contract: Contract,
-        entry_point: EntryPoint,
+        identifier: CallContractIdentifier,
+        entry_point_name: &str,
         args: RuntimeArgs,
     ) -> Result<CLValue, Error> {
+        let (contract, contract_hash, contract_package) = match identifier {
+            CallContractIdentifier::Contract { contract_hash } => {
```

Comment:
I tried different approaches, but I'm struggling to see an easy way to refactor this to avoid duplication nicely. The only thing that came to my mind was to use `self.context.read_gs_typed` rather than much longer `read_gs.` Hope that helps